### PR TITLE
Build fix

### PR DESCRIPTION
**Pull request by guygastineau**
This lets the package build for now ;)

It builds and all the tests pass.  The examples all work well.

I have a mix of `==` and `>=` as well as more explicit package versions in the `stack.yaml` than I expected to need... :/  I will try to clean it up tonight; if anyone has strong feelings about capturing the spirit of version bounds with `Cabal` please share your opinions for this PR ;)

I have neither changed information about the GHC version against which the tests were built (it was `GHC 8.10.2` with `base-4.13.0.0` BTW) nor bumped any internal versions for this package.  I imagine that some minor changes to this information must happen given the dependency chain I discovered when upgrading to patched `haskell-gi`.  So, I don't think this is ready for merging right now, but it should work as a patch to get projects building today.

@owickstrom from twitter it appears like you are taking some personal time (and perhaps you are a new father). CONGRATULATIONS! :tada:  Your relative absence recently is understandable :heart: So, just get to this whenever you can. (PS. Sorry for pinging you when it says `busy`... :/)

I will, for my part, keep this PR as a pointer to a building version of the project.  I encourage anyone experiencing difficulty building with `gi-gtk-declarative` to use my fork as a stop gap.  Hopefully, we can get a good patch put together, and it will be easy to merge into upstream here once the author is back.

### CODE REVIEW

**Comment by guygastineau** [comment]
I am not using `Nix`, so I don't know why the build failed.  If someone else wants to check that out, be my guest, but I think there have been problems with `Nix` and this package for a while, no?

**Comment by guygastineau** [comment]
Also, this PR is related to #89.  It should fix it, and it should be noted that I used information from the issue to create this patch ;), so  thank you to those who discussed workarounds in the issue!

**Comment by ptvirgo** [comment]
Thanks for putting this together.

**Comment by owickstrom** [comment]
Thanks for having a go at this and for the kind words. Appreciated!

Yeah, I don't have much time to dig deep here at the moment, but some quick feedback:

* Some version ranges in the Cabal file are replaced by exact versions, why is this? We should instead raise the upper bound.
* There are some new dependencies listed. I guess they might be required as extra-deps in Stack (I'm not sure at all!) but they shouldn't be in the Cabal file at all, if this package doesn't directly use them.
* I guess the Stack config is going to become problematic in the longer run. Is there not an LTS that would work for this package?

@cblp would very much appreciate some help reviewing this stuff if you have time. :eyes: 

Thanks again!

File: .gitignore
Comment (cblp) [inline .gitignore:12]: stack.yaml.lock helps to reproduce builds when packages on Hackage have small patches without version change. There is no need to ignore it.

**Comment by cblp** [comment]
There is some problem with nix:
```
error: anonymous function at /nix/store/9ndwxdkrjxhraw5ivg0rz8k7z35fhppr-cabal2nix-gi-gtk-declarative/default.nix:1:1 called without required argument 'gi-harfbuzz', at /nix/store/ssz8kir26kd6savyhq30vissmjlcnqpp-nixpkgs-unstable-2020-04-05/pkgs/development/haskell-modules/make-package-set.nix:87:27
```
Who knows nix and can help?

File: .gitignore
Comment (guygastineau) [reply .gitignore:12]: `stack.yaml.lock` wasn't under version control, and I was tired of avoiding `git add .`.  I am more than happy to drop the commit in a rebase before merging, but if the file is important I would prefer to include it in the version control.

**Comment by guygastineau** [comment]
@owickstrom the major cause for dependency inclusion was setting versions for the `examples/` projects to build.

I am sure, however, that my work here is sloppy and that the files have become verbose ;)  I will try to whittle the redundancy out of it.

Like I said in my commit message, I started using `==` just to try getting things to build.  Then I started using `>=`, so it will at least accept an upper bound later.  So, indeed, this is not ideal.  I will work on fixing it.  Your suggestion about finding the right LTS for `stack` is a good idea.  I find the build tooling to be the hardest part of Haskell (so much so that I have avoided using it in favor of Rust for projects at work where I would really prefer Haskell).  I am trying to get past that, so hopefully cleaning this up will help me get better at using both `stack and `cabal`.

File: .gitignore
Comment (cblp) [reply .gitignore:12]: OK, leave it ignored

**Comment by owickstrom** [comment]
@guygastineau Sounds good! Didn't mean to sound harsh or anything like that. Appreciate the work you've put into this, and I'm sure it'll be good. :slightly_smiling_face:

**Comment by owickstrom** [comment]
I can probably look at the nix stuff once the Cabal files are cleaned up. :+1:

**Comment by guygastineau** [comment]
Cool.  No worries; I didn't take offense.

I started work off with a bang this week, and I think I will actually have time to clean this up before the week is over ;)

**Comment by guygastineau** [comment]
FYI, for anyone following this.  I messed with the newest stack `lts`, and it is a bit easier, but there are still a lot of `gi-*` dependencies that need to be a higher version than the latest solver has (at least as far as I could tell).

Luckily, this gives me a way to clean up a bit, and I will be able to make the cabal constraints clearer and better.  However, the `stack` config is going to have more `extra-deps` for the time being at least than most of us would prefer for best practice.  Maybe the best course of action will be to keep things working in a convoluted stack config here until a new `lts` is released which fixes these instabilities in the ecosystem?  Otherwise we need to go out and increase upper bounds in related `gi-*` and `g*k-*` projects... maybe there aren't enough hands out there, so maybe those projects really do need people to help, IDK.

My main point here is, it seems like no solvers provided by stack can solve the dependencies for this project without a lot of `extra-deps` elements.  For the time being I will just try to refine that list against the latest resolver, `lts-16.27`, until such time as we may use a new solver that should include most of the package versions getting stuffed into `extra-deps` now.  So, it appears that some `uglier-than-preferrable` stop-gap measure is necessary for stack until a new resolver is released.

Please, if anyone can find some way that I am incorrect about this, LET ME KNOW!  I would love nothing more than to be wrong about my investigations this morning ;)

**Comment by guygastineau** [comment]
OK, now the cabal files have most of the version bounds removed.  Limiting `gi-gtk` to `<4` makes everything `just work` with `cabal v2-test` on each of the packages.  This seems better than before, but I imagine it is still better to have more version ranges defined?  IDK, I am still trying to figure out the right way to do things in the spirit of `cabal` the tool.

File: stack.yaml
Comment (guygastineau) [inline stack.yaml:1]: It won't build with the latest `lts-16.27`, but it builds and passes all tests on my machine with this nightly resolver.  I imagine this is not ideal (it is also using bleeding edge ghc, which I enjoy, but I guess it i not considered stable... :/).  Regardless, it seems fine as a way to give us the ability to keep using the project right now.

Comment (guygastineau) [reply stack.yaml:1]: I can try building with the `nightly-2020-12-14` resolver too.  It only requires `GHC 8.10.2` instead of `GHC 8.10.3`, but I don't think any earlier snapshots will work at all with the upstream `gi` fixes.

**Comment by guygastineau** [comment]
Travis is stuck.  Oh well, ;)

File: gi-gtk-declarative-app-simple/gi-gtk-declarative-app-simple.cabal
Comment (guygastineau) [inline gi-gtk-declarative-app-simple/gi-gtk-declarative-app-simple.cabal:36]: So, really when I look at the diff here with my cleaner changes it looks like I just need to extend the upper bounds.  I still want to increase the lower bounds too though, since it won't build with some of the current lower bounds.  I am not sure about conventions for that, but at least it should be easier to start from the current work (which builds) and start adding the bounds in again.

**Comment by Dretch** [comment]
I have taken the liberty of pushing some Nix changes to this branch (hope this is okay!):
1. I updated to a newer nixpkgs version: this appears to fix the build issues: `nix-build` now works.
2. I hacked `docs/requirements.nix` a bit to fix compatibility with the new nixpkgs.

Unfortunately the Travis build is probably going to fail because it goes over the 50 minute timeout. As discussed above it probably makes sense to migrate to Github actions anyway: this is very straightforward - I'll send another PR for it.

**Comment by guygastineau** [comment]
@Dretch great, thank you very much!

Haha, yeah that Travis log! It is just too much for that poor little VM to handle :cry: 

Okay, I will take a look at your other PR. Would we need to merge your CI fix before merging this one?
Oh how I look forward to having this merged and {H,St}ackage being updated ;)

**Comment by owickstrom** [comment]
The pull request for GitHub Actions CI is now merged. :tada: We'll see if I managed to get the auth tokens correct for Cachix.

**Comment by guygastineau** [comment]
So I just need to rebase this branch on top of the changes?

**Comment by guygastineau** [comment]
Okay, I rebased on the new master and force pushed to trigger the GA build.
@Dretch, thank you for your work with the `nix` stuff. Please not I force pushed the rebase, so your clone will be broken now.  I feel like that is so taboo, but it was the simplest thing to do here ;)

**Comment by guygastineau** [comment]
Watching the build log in real time I think it can be slimmed down some.

Do we need to run all the tests and generate all the docs for Codegen?

**Comment by guygastineau** [comment]
Alright, now it is just building the docs.  It looks like it will take over 60 minutes though.  Since the project only gets 2000 minutes a month I hope that doesn't cause problems... :/

Anyway, I am excited to have this passing the automated build now, but I am not sure if it is ready for merging.  It still relies on a nightly stack resolver for the stack config.  I don't know if this is a problem? Let me know you all think, and let me know if there are any other issues you can think up :)

**Comment by Dretch** [comment]
@guygastineau I just re-ran some builds to try and see if it got faster the second time... but I now realise it won't work (yet).

Nix (+Cachix) only caches stuff from builds on branches in the main repo (otherwise PRs could poison the cache), so this branch will always be slow. Once the branch is merged, though, the caching should work and the build will probably take 5 minutes or so. Future PRs will be able to read the cache, so they will be fast (unless they also change the nixpkgs version).

File: .gitignore
Comment (owickstrom) [reply .gitignore:12]: I know too little about Stack nowadays, but for lockfiles in general, they're useful for applications rather than libraries. If we can't disable it in the Stack configuration, I think it should be at least ignored.

File: stack.yaml
Comment (owickstrom) [reply stack.yaml:1]: Do you know why it doesn't build with the LTS?

I don't have any strong preferences here, I'm not using Stack so I can't say what would be best. But perhaps LTS + a few extra-deps is better than a nightly resolver?

**Comment by owickstrom** [comment]
Just to note: I've pushed a new change that uses the signing key instead of an API key for Cachix. Should just make the branch builds in this repo pass. PRs shouldn't push to Cachix, so it shouldn't make a difference for this PR.

**Comment by guygastineau** [comment]
@Dretch, thank you for the explanation.

File: stack.yaml
Comment (guygastineau) [reply stack.yaml:1]: Hmm, so I was expanding a project that uses Streamly and Evdev, and I noticed that for that project I needed to use the nightly resolver.  It was much easier for me to add 2 extra-deps that were older.  When I tried to build the project using `LTS` it went to dep hell, because some resources wanted older `gi-*` versions.

It might not matter if this library is built with `LTS` though (I can still try that if we want), but obviously there are big issues with using this along with `lts-16.27`.

At this point it seems to me like we should use the nightly resolver here.  It works the best, and it is a reminder to move to lts-16.x whenever the `gi-*` dependencies are up to date in an lts resolver.  Does that make sense to anyone else?

Comment (Dretch) [reply stack.yaml:1]: I have not tried it, but might `lts-17` (released a couple of days ago) work?

Comment (guygastineau) [reply stack.yaml:1]: Oh, I hadn't been paying attention to new lts releases. I expect it to work. I will try it out.  thank you!

**Comment by guygastineau** [comment]
I accidentally rebased on my out of date local master.  I am going to try out the newest stack LTS, so I will just push again once I have that tested.

**Comment by guygastineau** [comment]
As far as I can tell the CI runner spent that last 20 minutes doing nothing.  It finally started processing the libraries (which it said it was doing without any output for a while ;) thanks GitHub ).

Anyway, it worked locally for me, and I expect it to pass here even if it takes another hour here... :/

I really appreciate all the help along the way with this.  I think this is probably ready to merge now (once the CI build finishes), so whenever someone with write privilege wants to check things over again go for it.

File: stack.yaml
Comment (guygastineau) [reply stack.yaml]: Just to resolve things here.
https://github.com/owickstrom/gi-gtk-declarative/pull/90/commits/fe542ea38437380d00e965963d2880d5c7c169ee
switched to using `lts-17.0`, and it is great.  Thank you for the heads up about the new lts, @Dretch !

File: .gitignore
Comment (guygastineau) [reply .gitignore:12]: I looked for some ways to avoid the lock file with stack, but I couldn't find anything that would do it... :/

So I guess ignoring it will be the way forward.

**Review by owickstrom — APPROVED** [review]

**Comment by owickstrom** [comment]
Awesome that the new LTS worked. Great work @guygastineau!

**Comment by guygastineau** [comment]
You are welcome!

I wonder, have you bumped the Hackage meta-data?
I have never maintained a Hackage package, so I don't really know much about it.

**Comment by owickstrom** [comment]
You mean upload a new version? I haven't, but can maybe do tomorrow.

**Comment by guygastineau** [comment]
@owickstrom yeah basically.  I am under the impression that Hackage trustees for a package can bump meta-data information to give the project a new hash so cabal and stack will re-download it without needing to bump any real version numbers.  Of course, I guess the minor version, x.x.+1, can be bumped in this case.  It still seems strange since we only messed with dependency ranges and the build system.  Obviously it is your call.  Thank you.

**Pull request merged**
